# Task 4-6

In [1]:
import pandas as pd
import numpy as np
import os

### Step 3. import orders_products_combined

In [3]:
orders_products_combined =pd.read_pickle('d:/Instacart Basket Analysis_2024-01-02/02 Data/Prepared Data/orders_products_combined.pkl')

### Step 4. check the shape of imported dataframe

In [3]:
orders_products_combined.shape

(32434489, 9)

In [4]:
orders_products_combined.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_previous_order  product_id  add_to_cart_order  reordered  
0                        0.0         196                  1          0  
1                        0.0       14084                  2          0  
2                        0.0       12427                  3          0  
3                        0.0       26088                  4          0  
4                        0.0       26405                  5          0

### Step 5. Determine a suitable way to combine the orders_products_combined dataframe with your products data set. 

In [16]:
df_prods_cleaned=pd.read_csv(r'd:/Instacart Basket Analysis_2024-01-02/02 Data/Prepared Data/products_cleaned.csv', index_col=False)

In [5]:
df_prods_cleaned.shape

(49672, 6)

In [17]:
df_prods_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49672 entries, 0 to 49671
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Unnamed: 0     49672 non-null  int64  
 1   product_id     49672 non-null  int64  
 2   product_name   49672 non-null  object 
 3   aisle_id       49672 non-null  int64  
 4   department_id  49672 non-null  int64  
 5   prices         49672 non-null  float64
dtypes: float64(1), int64(4), object(1)
memory usage: 2.3+ MB


In [18]:
df_prods_cleaned.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [19]:
# for memory optimazing, changing the data type. 

df_prods_cleaned['product_id'] =df_prods_cleaned['product_id'].astype('int32')
df_prods_cleaned['aisle_id']=df_prods_cleaned['aisle_id'].astype('int8')
df_prods_cleaned['department_id']=df_prods_cleaned['department_id'].astype('int32')


In [20]:
# drop the Unnamed: 0 column
df_prods_cleaned = df_prods_cleaned.drop(columns ='Unnamed: 0')

In [21]:
df_prods_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49672 entries, 0 to 49671
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_id     49672 non-null  int32  
 1   product_name   49672 non-null  object 
 2   aisle_id       49672 non-null  int8   
 3   department_id  49672 non-null  int32  
 4   prices         49672 non-null  float64
dtypes: float64(1), int32(2), int8(1), object(1)
memory usage: 1.2+ MB


In [25]:
df_prods_cleaned.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

# merge two dataframe

In [23]:
ords_prods_merge = orders_products_combined.merge(df_prods_cleaned, on = 'product_id', indicator=True)

### Step 6. Confirm the results of the merge using the merge flag.

In [24]:
ords_prods_merge.head(20)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0    2539329        1             1                  2                  8   
1    2398795        1             2                  3                  7   
2     473747        1             3                  3                 12   
3    2254736        1             4                  4                  7   
4     431534        1             5                  4                 15   
5    3367565        1             6                  2                  7   
6     550135        1             7                  1                  9   
7    3108588        1             8                  1                 14   
8    2295261        1             9                  1                 16   
9    2550362        1            10                  4                  8   
10   2968173       15            15                  1                  9   
11   1870022       15            17                  2                 16   
12   1911383       15            18                  2                 11   
13   2715276       15            21                  1                  9   
14    487368       15            22                  1                 10   
15   2293453       19             2                  5                 14   
16   1973799       19             5                  6                 12   
17    532817       19             7                  4                 17   
18   1573906       21            10                  3                 10   
19   1593000       31            10                  3                  8   

    days_since_previous_order  product_id  add_to_cart_order  reordered  \
0                         0.0         196                  1          0   
1                        15.0         196                  1          1   
2                        21.0         196                  1          1   
3                        29.0         196                  1          1   
4                        28.0         196                  1          1   
5                        19.0         196                  1          1   
6                        20.0         196                  1          1   
7                        14.0         196                  2          1   
8                         0.0         196                  4          1   
9                        30.0         196                  1          1   
10                        7.0         196                  2          0   
11                        8.0         196                  6          1   
12                        7.0         196                  1          1   
13                        7.0         196                  1          1   
14                       14.0         196                  1          1   
15                        6.0         196                  3          0   
16                        8.0         196                 15          1   
17                        6.0         196                  1          1   
18                        6.0         196                  2          0   
19                        7.0         196                 17          0   

   product_name  aisle_id  department_id  prices _merge  
0          Soda        77              7     9.0   both  
1          Soda        77              7     9.0   both  
2          Soda        77              7     9.0   both  
3          Soda        77              7     9.0   both  
4          Soda        77              7     9.0   both  
5          Soda        77              7     9.0   both  
6          Soda        77              7     9.0   both  
7          Soda        77              7     9.0   both  
8          Soda        77              7     9.0   both  
9          Soda        77              7     9.0   both  
10         Soda        77              7     9.0   both  
11         Soda        77              7     9.0   both  
12         Soda        77              7     9.0   both 

In [27]:
ords_prods_merge.shape

(32404859, 14)

In [17]:
print(orders_products_combined.isnull().sum())

order_id                     0
user_id                      0
order_number                 0
order_day_of_week            0
order_hour_of_day            0
days_since_previous_order    0
product_id                   0
add_to_cart_order            0
reordered                    0
dtype: int64


In [20]:
ords_prods_merge['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

## ords_prods_merge.shape --> (32404859, 15), this is less then 32434489, so I use 'left_join'

# merge by 'how = left'

In [31]:
orders_products_combined.sort_index(inplace=True)

In [32]:
duplicate_indices = orders_products_combined.index[orders_products_combined.index.duplicated()]

In [33]:
orders_products_combined

order_id  user_id  order_number  order_day_of_week  \
0          2539329        1             1                  2   
1          2539329        1             1                  2   
2          2539329        1             1                  2   
3          2539329        1             1                  2   
4          2539329        1             1                  2   
...            ...      ...           ...                ...   
32434484   2977660   206209            13                  1   
32434485   2977660   206209            13                  1   
32434486   2977660   206209            13                  1   
32434487   2977660   206209            13                  1   
32434488   2977660   206209            13                  1   

          order_hour_of_day  days_since_previous_order  product_id  \
0                         8                        0.0         196   
1                         8                        0.0       14084   
2                         8                        0.0       12427   
3                         8                        0.0       26088   
4                         8                        0.0       26405   
...                     ...                        ...         ...   
32434484                 12                        7.0       14197   
32434485                 12                        7.0       38730   
32434486                 12                        7.0       31477   
32434487                 12                        7.0        6567   
32434488                 12                        7.0       22920   

          add_to_cart_order  reordered  
0                         1          0  
1                         2          0  
2                         3          0  
3                         4          0  
4                         5          0  
...                     ...        ...  
32434484                  5          1  
32434485                  6          0  
32434486                  7          0  
32434487                  8          0  
32434488                  9          0  

[32434489 rows x 9 columns]

In [34]:
df_prods_cleaned.sort_index(inplace=True)

In [36]:
duplicate_indices = df_prods_cleaned.index[df_prods_cleaned.index.duplicated()]


In [37]:
duplicate_indices


Index([], dtype='int64')

In [40]:
ords_prods_merge1 = pd.merge(orders_products_combined, df_prods_cleaned, on='product_id', how='left', indicator=True)

In [41]:
ords_prods_merge1['_merge'].value_counts()

_merge
both          32404859
left_only        30200
right_only           0
Name: count, dtype: int64

In [42]:
duplicate_records = orders_products_combined[orders_products_combined.duplicated(subset='product_id', keep=False)]


In [43]:
duplicate_records


order_id  user_id  order_number  order_day_of_week  \
0          2539329        1             1                  2   
1          2539329        1             1                  2   
2          2539329        1             1                  2   
3          2539329        1             1                  2   
4          2539329        1             1                  2   
...            ...      ...           ...                ...   
32434484   2977660   206209            13                  1   
32434485   2977660   206209            13                  1   
32434486   2977660   206209            13                  1   
32434487   2977660   206209            13                  1   
32434488   2977660   206209            13                  1   

          order_hour_of_day  days_since_previous_order  product_id  \
0                         8                        0.0         196   
1                         8                        0.0       14084   
2                         8                        0.0       12427   
3                         8                        0.0       26088   
4                         8                        0.0       26405   
...                     ...                        ...         ...   
32434484                 12                        7.0       14197   
32434485                 12                        7.0       38730   
32434486                 12                        7.0       31477   
32434487                 12                        7.0        6567   
32434488                 12                        7.0       22920   

          add_to_cart_order  reordered  
0                         1          0  
1                         2          0  
2                         3          0  
3                         4          0  
4                         5          0  
...                     ...        ...  
32434484                  5          1  
32434485                  6          0  
32434486                  7          0  
32434487                  8          0  
32434488                  9          0  

[32434358 rows x 9 columns]

In [13]:
duplicate_values = df_prods_cleaned[df_prods_cleaned.duplicated(subset='product_id', keep=False)]

In [14]:
duplicate_values

product_id                                       product_name  \
6784         6800                          Revive Zero Vitamin Water   
6785         6800                 Sprouted Quinoa Flakes Baby Cereal   
26504       26520  Clinical Advanced Solid Ultimate Fresh Anti-Pe...   
26505       26520       Cheese Shredded Sharp Cheddar Reduced Fat 2%   

       aisle_id  department_id  prices  
6784         64              7       6  
6785         92             18      14  
26504        80             11      10  
26505        21             16       2

# duplicated values were deleted at previous exercise 4-5, but those are the records that all records are same in every columns.  Still, There are 2 product_ids having 2 different product names.  

In [15]:
# if product_id ordered by department, we can find that which product rellay has product ID 6800. 
# e.g. if 6800 ~ 6900 are beverage, then 6800 should be Revive Zero Vitamin Water 
df_prods_cleaned[(df_prods_cleaned['product_id'] > 6790) & (df_prods_cleaned['product_id'] <6810)]


product_id                                       product_name  aisle_id  \
6776        6791                 Low Fat Blueberry Goat Milk Yogurt       120   
6777        6792             French Style Green Beans No Salt Added        81   
6778        6793                                  Mozzarella Sticks      -127   
6779        6794                                Thin Regular Liners       126   
6780        6795                            Organic Garlic Granules       104   
6781        6796   Restaurante Style Sea Salt & Lime Tortilla Chips       107   
6782        6797                             Boneless Pork Shoulder       122   
6783        6798          18-in-1 Hemp Eucalyptus Pure Castile Soap        25   
6784        6800                          Revive Zero Vitamin Water        64   
6785        6800                 Sprouted Quinoa Flakes Baby Cereal        92   
6786        6801    Family Pack Bites Juicy Red & Green Apple Mixed       123   
6787        6802  Toothpaste for Sensitive Teeth and Cavity Prot...        20   
6788        6803                                      Rose Lemonade        98   
6789        6804         Natural Hawaiian Facial Mask Papaya Enzyme       109   
6790        6805                              Raspberry Ginger Brew        77   
6791        6806                        Iced Gingerbread Energy Bar         3   
6792        6807                                      Pancake Syrup        29   
6793        6808                Organic Spinach & Cheese Tortellini        12   
6794        6809                       Berry Bunnies Cereal Organic       121   

      department_id  prices  
6776             16       6  
6777             15      10  
6778              1       6  
6779             11       5  
6780             13       9  
6781             19       4  
6782             12      19  
6783             11       9  
6784              7       6  
6785             18      14  
6786              4       9  
6787             11       8  
6788              7      10  
6789             11      15  
6790              7       3  
6791             19       5  
6792             13      11  
6793              9      13  
6794             14       7

In [16]:
df_prods_cleaned[(df_prods_cleaned['product_name'] == 'Sprouted Quinoa Flakes Baby Cereal')]


product_id                        product_name  aisle_id  department_id  \
6785        6800  Sprouted Quinoa Flakes Baby Cereal        92             18   

      prices  
6785      14

In [17]:
df_prods_cleaned[(df_prods_cleaned['product_name'] == 'Revive Zero Vitamin Water')]


product_id               product_name  aisle_id  department_id  prices
6784        6800  Revive Zero Vitamin Water        64              7       6

In [18]:
# is there any duplicated product_name?

duplicate_names = df_prods_cleaned[df_prods_cleaned.duplicated(subset='product_name', keep=False)]

In [19]:
duplicate_names

Empty DataFrame
Columns: [product_id, product_name, aisle_id, department_id, prices]
Index: []

In [20]:
# there is no duplicated product name. 

# If 1 out of 2 product_id 6800 and 26520 are deleted, how the value count will be changed with left merge?

In [22]:
# delete product_id 6800 and 26520
df_prods_cleaned_v2 = df_prods_cleaned.drop_duplicates(subset='product_id', keep='first')

In [23]:
df_prods_cleaned_v2

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49667       49684          Vodka, Triple Distilled, Twist of Vanilla   
49668       49685                 En Croute Roast Hazelnut Cranberry   
49669       49686                                   Artisan Baguette   
49670       49687         Smartblend Healthy Metabolism Dry Cat Food   
49671       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19       5  
1           104             13       9  
2            94              7       4  
3            38              1      10  
4             5             13       4  
...         ...            ...     ...  
49667       124              5       5  
49668        42              1       3  
49669       112              3       7  
49670        41              8       4  
49671        73             11      13  

[49670 rows x 5 columns]

In [24]:
df_prods_cleaned_v2.shape

(49670, 5)

# left merge with second cleaned dataframe

In [25]:
ords_prods_merge_v2 = pd.merge(orders_products_combined, df_prods_cleaned_v2, on='product_id', how='left', indicator=True)

In [26]:
ords_prods_merge_v2['_merge'].value_counts()

_merge
both          32404289
left_only        30200
right_only           0
Name: count, dtype: int64

In [27]:
print(32404289+30200)

32434489


# finally the result is same as 32434489!!

## But so far, I don't know which product name is correct 6800 and 26520, and I cannot find any clue for that. 

## The gap between 32434489 (before merging) and 32404859 (inner merging) is only 29,630 (less than 0.1%), I adopt 32404859 (inner merging)

In [ ]:
# export ords_prods_merge
ords_prods_merge.to_pickle('d:/Instacart Basket Analysis_2024-01-02/02 Data/Prepared Data/ords_prods_merge.pkl')